src/surfaceqg.jl

```diff
@@ -0,0 +1,370 @@
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh = 2*grid.Krsq^params.nν * abs2(sol)
+  CUDA.@allowscalar vars.uh[1, 1] = 0
+  return params.ν / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+"""
+    buoyancy_work(prob)
+    buoyancy_work(sol, v, grid)
+
+Returns the domain-averaged rate of work of buoyancy variance by the forcing Fh.
+"""
+@inline function buoyancy_work(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh =  sol * conj(vars.Fh)
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+@inline function buoyancy_work(sol, vars::StochasticForcedVars, grid)
+  @. vars.uh =  (vars.prevsol + sol) / 2 * conj(vars.Fh) # Stratonovich
+  # @. vars.uh = grid.invKrsq * vars.prevsol * conj(vars.Fh)           # Ito
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+@inline work(prob) = work(prob.sol, prob.vars, prob.grid)
+
+"""
+    buoyancy_drag(prob)
+
+Returns the extraction of domain-averaged buoyancy variance by drag/hypodrag μ.
+"""
+@inline function buoyancy_drag(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+  @. vars.uh = grid.Krsq^params.nμ * abs2(sol)
+  vars.uh[1, 1] = 0
+  return params.μ / (grid.Lx * grid.Ly) * parsevalsum(vars.uh, grid)
+end
+
+"""
+    buoyancy_advection(prob)
+
+Returns the average of domain-averaged advection of buoyancy variance. This should
+be zero if buoyancy variance is conserved.
+"""
+@inline function buoyancy_advection(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+
+  @. vars.bh = sol
+  @. vars.uh =   im * grid.l  * sqrt.(grid.invKrsq) * sol
+  @. vars.vh = - im * grid.kr * sqrt.(grid.invKrsq) * sol
+
+  ldiv!(vars.u, grid.rfftplan, vars.uh)
+  ldiv!(vars.v, grid.rfftplan, vars.vh)
+  ldiv!(vars.b, grid.rfftplan, vars.bh)
+
+  @. vars.u *= vars.b # u*b
+  @. vars.v *= vars.b # v*b
+
+  mul!(vars.uh, grid.rfftplan, vars.u) # \hat{u*b}
+  mul!(vars.vh, grid.rfftplan, vars.v) # \hat{v*b}
+
+  @. vars.bh = ( - im * grid.kr * vars.uh - im * grid.l * vars.vh ) * conj( vars.bh )
+
+  return 1 / (grid.Lx * grid.Ly) * parsevalsum(vars.bh, grid)
+end
+
+"""
+    kinetic_energy_advection(prob)
+
+Returns the average of domain-averaged advection of kinetic energy by the
+leading-order (geostrophic) flow.
+"""
+@inline function kinetic_energy_advection(prob)
+  sol, vars, params, grid = prob.sol, prob.vars, prob.params, prob.grid
+
+  @. vars.b = vars.u * vars.u
+  mul!(vars.bh, grid.rfftplan, vars.u)
+  sol = vars.bh
+  @. vars.v = vars.v * vars.u
+  mul!(vars.bh, grid.rfftplan, vars.u)
+  @. vars.uh = - ( im * grid.kr * sol + im * grid.l * vars.bh ) * conj(vars.uh)
+
+  @. vars.u = vars.u * vars.v
+  mul!(vars.bh, grid.rfftplan, vars.u)
+  sol = vars.bh
+  @. vars.v = vars.v * vars.v
+  mul!(vars.bh, grid.rfftplan, vars.v)
+  @. vars.vh = - ( im * grid.kr * sol + im * grid.l * vars.bh ) * conj(vars.vh)
```

Comment:
What's going on here? It looks like there's some finagling with temporary variables that leads to possibly confusing expressions. Comments could help. Another option is to use different names. For example we could use something like

```julia
u² = vars.b
@. u² = vars.u^2

vars.v .*= vars.u
uv = vars.v
```

et cetera.

are perfectly fine.

Reply:
Doesn't this induce allocating new arrays in memory @glwagner ?

Reply:
No.

```julia
julia> a = rand(1024, 1024);

julia> @time for i = 1:100
       b = a
       end
  0.000000 seconds

julia> @time for i = 1:100
       c = deepcopy(a)
       end
  0.270146 seconds (400 allocations: 800.043 MiB, 39.13% gc time)
```

The equals sign (`=`) performs a name binding --- in other words `b=a` when `a` is an array means "bind the name `b` to the memory in `a`). Then `b[1, 1]` accesses the _same_ memory as `a[1, 1]`. `b=a` does not copy an array. For that, you must use `copy` or `deepcopy`.

Reply:
If there is an _expression_ on the right side of the equals sign that requires memory allocation, then a new array will be created for that purpose, and the memory associated with that new array will be bounded to the name on the left side of the equal sign. This is what occurs when you write `b = a .* c`.

Reply:
The classic demonstration of my first comment is

```julia
julia> a = rand(2, 2)
2×2 Array{Float64,2}:
 0.442335  0.153258
 0.377039  0.558756

julia> b = a
2×2 Array{Float64,2}:
 0.442335  0.153258
 0.377039  0.558756

julia> a[1, 1] = 0
0

julia> b
2×2 Array{Float64,2}:
 0.0       0.153258
 0.377039  0.558756
```

Reply:
dang! of course..

(these sort of name-bindings can help clarify code in other modules btw..)